reco_utils/dataset/movielens.py

```diff
@@ -377,48 +424,60 @@ def _get_schema(header, schema):
+    """Download and extract MovieLens rating and item datafiles if they don’t already exist.
 
+    Args:
+        size (str): Size of the data to load. One of ("100k", "1m", "10m", "20m").
+        local_cache_path (str): Path where to cache the zip file locally.
+            If None, use default, which is "data/ml-{size}/ml.zip"
+
+    Returns:
+        str: rating datafile path,
+        str: item datafile path.
+    """
     if size not in DATA_FORMAT:
         raise ValueError(ERROR_MOVIE_LENS_SIZE)
-    if not local_cache_path.endswith(".zip"):
+    if local_cache_path is None:
+        default_dir = os.path.join("data", "ml-{}".format(size))
+        local_cache_path = os.path.join(default_dir, "ml.zip")
+    elif not local_cache_path.endswith(".zip"):
         raise ValueError(ERROR_LOCAL_CACHE_PATH)
 
     path, filename = os.path.split(os.path.realpath(local_cache_path))
-
-    # Make sure a temporal zip file get cleaned up no matter what
-    atexit.register(_clean_up, local_cache_path)
-
-    maybe_download(
-        "http://files.grouplens.org/datasets/movielens/ml-" + size + ".zip",
-        filename,
-        work_directory=path,
-    )
+    if not os.path.exists(path):
+        os.makedirs(path)
 
     _, dataname = os.path.split(DATA_FORMAT[size].path)
     datapath = os.path.join(path, dataname)
     _, item_dataname = os.path.split(DATA_FORMAT[size].item_path)
     item_datapath = os.path.join(path, item_dataname)
 
-    with ZipFile(local_cache_path, "r") as z:
-        with z.open(DATA_FORMAT[size].path) as zf, open(datapath, 'wb') as f:
-            shutil.copyfileobj(zf, f)
-        with z.open(DATA_FORMAT[size].item_path) as zf, open(item_datapath, 'wb') as f:
-            shutil.copyfileobj(zf, f)
+    if not os.path.exists(dataname) or not os.path.exists(item_datapath):
+        download_movielens(size, filename, path)
+        _extract(size, local_cache_path, datapath, item_datapath)
+        # Remove downloaded zip file since we don't need anymore.
+        os.remove(local_cache_path)
 
-    _clean_up(local_cache_path)
+    return datapath, item_datapath
 
-    # Make sure a temporal data file get cleaned up when done
-    atexit.register(_clean_up, datapath)
-    atexit.register(_clean_up, item_datapath)
 
-    return datapath, item_datapath
+def download_movielens(size, dest_filename, dest_path="."):
+    """Download MovieLens datafile.
+
+    Args:
+        size (str): Size of the data to load. One of ("100k", "1m", "10m", "20m").
+        dest_filename (str): zip filename
+        dest_path (str): Work directory
+    """
+    url = "http://files.grouplens.org/datasets/movielens/ml-" + size + ".zip"
+    maybe_download(url, dest_filename, work_directory=dest_path)
 
 
-def _clean_up(filepath):
-    """ Remove cached file. Be careful not to erase anything else. """
-    try:
-        os.remove(filepath)
-    except OSError:
-        pass
+def _extract(size, local_cache_path, datapath, item_datapath):
```

Comment:
this is pretty specialized for movielens right? I don't think we need to separate it out as its own function

Reply:
Same reason here, I think we need something like this. I would even do it public, this is something that we could use independently

Reply:
Right. This was used from multiple functions before the refactoring, which is replaced by _maybe_download_and_extract.

Reply:
Ok, github is not loading the latest thread automatically, so didn't see @miguelgfierro's comment when I replied, and this makes the discussion really complicated :-(

I prefer to merge this `_extract` function into `_maybe_download_and_extract`, because as Scott pointed out, extract will work only after the zip file is downloaded. Regarding the single responsibility, I guess the function is small enough and its name describes the task enough. A similar example is from tensorflow's tutorial's [cifar10 dataset downloader](https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10.py) which has maybe_download_and_extract function.

Regarding `download_movielens`, I think we may want to keep it separately. As Miguel pointed out, someone may want to just download it so that he can use other files like tag or user info (100k data includes user info). Currently, `_maybe_download_and_extract` only extracts rating and item data. One good reason for having this download wrapper is, to make sure the users download the same data as the repo's data-loader.

Reply:
I think maybe_download_and_extract is ok, however, I ended up changing that. Please take a look at this function: https://github.com/Microsoft/Recommenders/blob/a66ec60b0fdd8be9e593dc22d94a8de98ee85df1/reco_utils/dataset/criteo_dac.py#L48

Basically there is something called load_criteo() which will do the whole thing, which is download, extract and load criteo. I created 3 steps if you look at the function. Initially, I thought of using download_and_extract, but I don't see the value of having 2 of the 3 steps together and then the load in a different steps. I think, at least in this case, it's clearer to have 3 single steps: download, extract and load. 

But again, I don't think it is super key as long as we follow single responsibility